File: Graph/Prims.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;

public class Prims {
    static class Edge {
        int src;
        int dest;
        int weight;

        public Edge(int src, int dest, int weight) {
            this.src = src;
            this.dest = dest;
            this.weight = weight;
        }
    }

    static class Pair implements Comparable<Pair> {
        int node;
        int cost;

        public Pair(int node, int cost) {
            this.node = node;
            this.cost = cost;
        }

        public int compareTo(Pair p2) {
            return this.cost - p2.cost;
        }
    }

    public static void createGraph(ArrayList<Edge> graph[]) {
        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<>();
        }

        graph[0].add(new Edge(0, 1, 10));
        graph[0].add(new Edge(0, 2, 6));
        graph[0].add(new Edge(0, 3, 5));

        graph[1].add(new Edge(1, 0, 10));
        graph[1].add(new Edge(1, 3, 15));
        graph[1].add(new Edge(1, 4, 8));

        graph[2].add(new Edge(2, 0, 6));
        graph[2].add(new Edge(2, 3, 4));
        graph[2].add(new Edge(2, 6, 6));

        graph[3].add(new Edge(3, 0, 5));
        graph[3].add(new Edge(3, 1, 15));
        graph[3].add(new Edge(3, 2, 4));

        graph[4].add(new Edge(4, 1, 8));
        graph[4].add(new Edge(4, 6, 3));

        graph[5].add(new Edge(5, 6, 4));
        graph[5].add(new Edge(5, 7, 7));
        graph[5].add(new Edge(5, 8, 6));

        graph[6].add(new Edge(6, 2, 6));
        graph[6].add(new Edge(6, 4, 3));
        graph[6].add(new Edge(6, 5, 4));

        graph[7].add(new Edge(7, 5, 7));

        graph[8].add(new Edge(8, 5, 6));
    }

    public static void primsAlgo(ArrayList<Edge> graph[], int V) {
        PriorityQueue<Pair> pq = new PriorityQueue<>();
        boolean[] vis = new boolean[V];
        int[] parent = new int[V];
        int[] key = new int[V];

        Arrays.fill(key, Integer.MAX_VALUE);
        Arrays.fill(parent, -1);
        key[0] = 0;

        pq.add(new Pair(0, 0));
        int mstCost = 0;

        while (!pq.isEmpty()) {
            Pair curr = pq.remove();

            if (!vis[curr.node]) {
                vis[curr.node] = true;
                mstCost += curr.cost;

                for (Edge e : graph[curr.node]) {
                    if (!vis[e.dest] && e.weight < key[e.dest]) {
                        key[e.dest] = e.weight;
                        parent[e.dest] = curr.node;
                        pq.add(new Pair(e.dest, e.weight));
                    }
                }
            }
        }

        System.out.println("Edge\tWeight");
        for (int i = 1; i < V; i++) {
            System.out.println(parent[i] + " - " + i + "\t" + key[i]);
        }
        System.out.println("Total Weight of MST: " + mstCost);
    }

    public static void main(String[] args) {
        int V = 9;
        ArrayList<Edge>[] graph = new ArrayList[V];
        createGraph(graph);
        primsAlgo(graph, V);
    }
}

File: Graph/Prims.cpp
#include <iostream>
#include <climits>
using namespace std;

struct Edge {
    int src;
    int dest;
    int weight;
    
    Edge() {}
    Edge(int src, int dest, int weight) {
        this->src = src;
        this->dest = dest;
        this->weight = weight;
    }
};

class Prims {
private:
    static const int MAX_VERTICES = 10;
    static const int MAX_EDGES_PER_VERTEX = 10;
    
    Edge graph[MAX_VERTICES][MAX_EDGES_PER_VERTEX];
    int edgeCount[MAX_VERTICES];
    
public:
    Prims() {
        for (int i = 0; i < MAX_VERTICES; i++) {
            edgeCount[i] = 0;
        }
    }
    
    void addEdge(int src, int dest, int weight) {
        if (edgeCount[src] < MAX_EDGES_PER_VERTEX) {
            graph[src][edgeCount[src]] = Edge(src, dest, weight);
            edgeCount[src]++;
        }
    }
    
    void createGraph() {
        addEdge(0, 1, 10);
        addEdge(0, 2, 6);
        addEdge(0, 3, 5);
        
        addEdge(1, 0, 10);
        addEdge(1, 3, 15);
        addEdge(1, 4, 8);
        
        addEdge(2, 0, 6);
        addEdge(2, 3, 4);
        addEdge(2, 6, 6);
        
        addEdge(3, 0, 5);
        addEdge(3, 1, 15);
        addEdge(3, 2, 4);
        
        addEdge(4, 1, 8);
        addEdge(4, 6, 3);
        
        addEdge(5, 6, 4);
        addEdge(5, 7, 7);
        addEdge(5, 8, 6);
        
        addEdge(6, 2, 6);
        addEdge(6, 4, 3);
        addEdge(6, 5, 4);
        
        addEdge(7, 5, 7);
        
        addEdge(8, 5, 6);
    }
    
    int findMinKey(int key[], bool vis[], int V) {
        int min = INT_MAX;
        int minIndex = -1;
        
        for (int v = 0; v < V; v++) {
            if (!vis[v] && key[v] < min) {
                min = key[v];
                minIndex = v;
            }
        }
        return minIndex;
    }
    
    void primsAlgo(int V) {
        bool vis[MAX_VERTICES];
        int parent[MAX_VERTICES];
        int key[MAX_VERTICES];
        
        // Initialize arrays
        for (int i = 0; i < V; i++) {
            vis[i] = false;
            parent[i] = -1;
            key[i] = INT_MAX;
        }
        
        key[0] = 0;
        int mstCost = 0;
        
        for (int count = 0; count < V; count++) {
            int u = findMinKey(key, vis, V);
            
            if (u == -1) break;
            
            vis[u] = true;
            mstCost += key[u];
            
            // Update key values of adjacent vertices
            for (int i = 0; i < edgeCount[u]; i++) {
                Edge e = graph[u][i];
                int v = e.dest;
                
                if (!vis[v] && e.weight < key[v]) {
                    key[v] = e.weight;
                    parent[v] = u;
                }
            }
        }
        
        cout << "Edge\tWeight" << endl;
        for (int i = 1; i < V; i++) {
            if (parent[i] != -1) {
                cout << parent[i] << " - " << i << "\t" << key[i] << endl;
            }
        }
        cout << "Total Weight of MST: " << mstCost << endl;
    }
};

int main() {
    int V = 9;
    Prims prim;
    prim.createGraph();
    prim.primsAlgo(V);
    
    return 0;
}
